debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/CellData.java

```diff
@@ -66,10 +67,33 @@ public boolean isPrimary() {
 
     @Override
     public Struct record(Schema schema) {
-        return new Struct(schema)
-                .put(CELL_VALUE_KEY, value)
+        Struct cellStruct = new Struct(schema)
                 .put(CELL_DELETION_TS_KEY, deletionTs)
                 .put(CELL_SET_KEY, true);
+
+        Schema valueSchema = schema.field(CELL_VALUE_KEY).schema();
+        if (valueSchema.type() == Schema.Type.STRUCT) {
+            Struct clonedValue = cloneValue(valueSchema, value);
+            cellStruct.put(CELL_VALUE_KEY, clonedValue);
+        }
+        else {
+            cellStruct.put(CELL_VALUE_KEY, value);
+        }
+
+        return cellStruct;
+    }
+
+    // Encountered DataException("Struct schemas do not match.") when cell value is a Struct,
+    // because equal() for Kafka Connect Schema is using the default shallow comparison.
+    // Replace the schema of value struct with the same schema object passed in to avoid this problem.
+    private Struct cloneValue(Schema valueSchema, Object value) {
+        Struct valueStruct = (Struct) value;
```

Comment:
Better declare `value` a `Struct` already, pushing the cast to the method call site, where it's known to be a struct.

Reply:
Could we also encounter a struct of struct, i.e. does this method have to be called recursively?

Reply:
> Better declare `value` a `Struct` already, pushing the cast to the method call site, where it's known to be a struct.

Updated accordingly :)

Reply:
> Could we also encounter a struct of struct, i.e. does this method have to be called recursively?

Nice catch! I tried to process a nested UDT data in cassandra and I was wrong about the cause of `DataException("Struct schemas do not match.")` here. Kafka Connect Schema does have an implemented equal() to compare schemas, the error is instead because the `valueSchema` is `optional`, but the schema of `value` formed during deserialization is not. Since the ultimate fix for this might touch a lot of places, I keep cloneValue() here for a temporary fix and created https://issues.redhat.com/browse/DBZ-2103 separately for this issue.

Reply:
Gotcha, makes sense!